### PR TITLE
[AIRFLOW-6790] Add basic Tableau Integration

### PR DESCRIPTION
**Pull request by feluelle**
This PR adds a **TableauHook** which wraps the tableauserverclient python library in an AirflowHook to connect to a Tableau Server. It also contains a **TableauRefreshWorkbookOperator** which refreshes a Tableau workbook / extract and a **TableauJobStatusSensor** which pokes the status of a Tableau Job until it has finished.

---
Issue link: [AIRFLOW-6790](https://issues.apache.org/jira/browse/AIRFLOW-6790)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Review by feluelle — COMMENTED** [review]
Tableau does belong to Salesforce but can be used separately.

So the question is whether we want to install it separately or together with `simple-salesforce`.

WDYT?

File: airflow/providers/salesforce/hooks/tableau.py
Comment (eladkal) [inline airflow/providers/salesforce/hooks/tableau.py]: Shouldn't `tableau_default` be created in `airflow/utils/db.py` ?

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py]: Wouldn't it be better default to be `None`?

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py:54]: Do you think it would be possible to have Tableau connection defined in `airflow/models/connection.py`
By having that the Extra field could have default site hence it would be able to extract the site from the connection rather than pass it every time the hook is used.
The hook itself can overwrite the default site in the connection.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py:92]: Does this support refreshing individual workbook report or the whole workbook ?
Say I have:
`https://url/#/site/site_name/views/Calls/Department`
`https://url/#/site/site_name/views/Calls/Agents`
if I pass `Calls` would it refresh both `Department` & `Agents` ?
What if I pass `Calls/Agents` ?

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py]: An empty string is the default value in the library. See https://tableau.github.io/server-client-python/docs/api-ref#authentication

File: airflow/providers/salesforce/hooks/tableau.py
Comment (feluelle) [reply airflow/providers/salesforce/hooks/tableau.py]: Actually I think I will remove the default. WDYT?
```suggestion
    def __init__(self, site_id='', tableau_conn_id):
```

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py:54]: Good point.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py:92]: I am gonna check that.

File: airflow/providers/salesforce/hooks/tableau.py
Comment (eladkal) [reply airflow/providers/salesforce/hooks/tableau.py]: I think it should have default. It can be useful for example dags. Especially if this connection will require settings in the Extra field (like discussed on the default site)

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py:92]: > Refreshes the specified workbook.
> 
> This method refreshes the specified workbook, with no need to associate the workbook refresh with a scheduled task. This method is the equivalent of selecting a workbook using the Tableau Server UI, and then selecting Refresh Extracts from the menu (also known as a "manual refresh").
> 

from https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_workbooksviews.htm#update_workbook_now

Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py:92]: > if I pass Calls would it refresh both Department & Agents ?

Yes.

> What if I pass Calls/Agents ?

I think it fails, because that's not a Workbook - it is a View in a Workbook, right?!

Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py:92]: You actually pass an ID of a workbook not the name to the Tableau Server. I am iterating through workbooks by names and I am returning the workbook id of the workbook that matches. -> which means it will not find `Calls/Agents`

**Comment by feluelle** [comment]
Ready for another review @eladkal :)

File: setup.py
Comment (Fokko) [inline setup.py]: Should we pin this version?

File: setup.py
Comment (Fokko) [inline setup.py]: I don't think Salesforce and Tableau are tightly integrated. I would split this into another group.

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (Fokko) [inline airflow/providers/salesforce/sensors/tableau_job_status.py]: Can we make a constant out of the `-1`? To avoid magic numbers in the code.

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (Fokko) [inline airflow/providers/salesforce/sensors/tableau_job_status.py]: I would move this one above the `**kwargs`

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (Fokko) [inline airflow/providers/salesforce/sensors/tableau_job_status.py]:
```suggestion
    def poke(self, context: dict) -> bool:
```
Type annotations ftw

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (Fokko) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py]:
```suggestion
    def _refresh_workbook(self, tableau_hook: TableauHook, workbook_id: str) -> str:
```
We all ❤️ type annotations.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (Fokko) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py]:
```suggestion
    def _get_workbook_by_name(self, tableau_hook: TableauHook) -> Workbook:
```

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (Fokko) [inline airflow/providers/salesforce/operators/tableau_refresh_workbook.py]:
```suggestion
    def execute(self, context: dict) -> str:
```

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (Fokko) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: Personally I'm not a big fan of this pattern. I know that this was introduced with the AWS stuff a while ago, but I'd like to have the original Operator to be blocking (maybe make this configurable), instead of always using an Operator/Sensor pair.

For example, if the `TableauRefreshWorkbookOperator` operation fails for some reason. We'd like to retry this. If the exit code is given at the Sensor, the sensor will be retried.

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (Fokko) [reply airflow/providers/salesforce/sensors/tableau_job_status.py]: What are the options in terms of return codes here? Can we check if the notebook successfully refreshed, or maybe if it threw an error while refreshing?

**Review by Fokko — COMMENTED** [review]
Nice work @feluelle. Having a Tableau operator would be a great addition. I've added some minor comments.

File: setup.py
Comment (feluelle) [reply setup.py]: I think we could (and maybe should) since the current version v0.9 seems to be near a v1 release which could be a release with breaking changes (I don't know). I couldn't find anything related to release schedule / plan nor versioning of releases.

Actually because you are bringing this up I am wondering if we should create general rules for pinning versions of dependencies. **This of course should be discussed in the dev list.**
It is a very complicated topic.

I think there are several points that could be relevant:
* knowing the releases and versioning of the dependency
* having one or more users in the community which have responsibilities in the project the dependency belongs to. (like committer status for apache projects, etc.)
* ...

If we have an issue due to a new release:
Maybe we should somewhere write down our dependencies and keep track of the issues we have towards non-pinning a dependency. So that we can better understand their releases. Every dependency has their own semantic versioning of releases.

Maybe we should add a list of requirements when we want to add a non-pinned dependency?!

File: setup.py
Comment (feluelle) [reply setup.py]: Okay. I am fine with this :)

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (feluelle) [reply airflow/providers/salesforce/sensors/tableau_job_status.py]: > Can we make a constant out of the -1? To avoid magic numbers in the code.

Good Point. I will do that.

> What are the options in terms of return codes here? Can we check if the notebook successfully refreshed, or maybe if it threw an error while refreshing?

Yes I need to check that! Thanks.
_The finishCode indicates the status of the job: 0 for success, 1 for error, or 2 for cancelled._ - REST API docs (not python client docs)
That is really weird, because I got a `-1` for a failed Job. I definitly need to check that again!

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (feluelle) [reply airflow/providers/salesforce/sensors/tableau_job_status.py]: Will do.

File: airflow/providers/salesforce/sensors/tableau_job_status.py
Comment (feluelle) [reply airflow/providers/salesforce/sensors/tableau_job_status.py]: Yes 👍 Thanks.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py]: Yes 👍 Thanks.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py]: Yes 👍 Thanks.

File: airflow/providers/salesforce/operators/tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/operators/tableau_refresh_workbook.py]: Yes 👍 Thanks.

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: > instead of always using an Operator/Sensor pair.

That's where I am using subdags :) ❤️

Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: >  I'd like to have the original Operator to be blocking (maybe make this configurable)

So you want it to `sleep` ? An extract refresh can take very long.

Comment (Fokko) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: Normally the Operator itself would block until it has refreshed. So the behavior of the Sensor is part of the Operator.
For me, having a separate sensor is a bit awkward. For example with the BashOperator, we wait until the process has finished, we don't have a sensor that will block until the process has ended.

File: setup.py
Comment (Fokko) [reply setup.py]: I think we should have this discussion indeed. However, for the Tableau operator, we know that it works for a specific version. Pinning that version now will ensure that it will work in the future.

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I could make this configurable so that after the refresh operation has been kicked off It will poke until it succeeds and then the task status will be success.
But this would mean that you can have this bad case where the refresh operation has been kicked off successfully but somehow the poke fails due to a unusual error. When you then clear this task (because it failed) you will kick off another extract refresh.

File: setup.py
Comment (feluelle) [reply setup.py]: If I put it into a separate group, does it mean it also belongs to the tableau provider? No, right?

File: setup.py
Comment (feluelle) [reply setup.py]: Yes, will do that.

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (eladkal) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I think having a separate sensor or making it configurable is crucial.

When you perform Extract the processing of the job is "moving" from Airflow worker to Tableau server. There is no reason for the Airflow operator to continue occupying the worker needlessly.

BashOperator is different. The script runs on the Airflow worker resources. It can also be used to connect with another service resulted in waiting time but this is an option unlike the TableauRefreshWorkbookOperator where it always executed on the Tableau server resources.

Think also about a case where many Dags uses TableauRefreshWorkbookOperator at the same time. If all occupy the workers till extract is finished it might cause the entire Airflow cluster to be paralyzed.

I think @feluelle concern about 2nd refresh is a valid one.

Comment (Fokko) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: That depends on how you implement it. If you fetch the and check if it is still refreshing, then you should not re-refresh the workbook. You need to handle different exceptions in a different way.

The implementation of the Operator is a different discussion. Right now this is getting cluttered in the codebase. For GCP we're having a single operator that blocks, while many AWS operators have this Operator/Sensor pattern.

Comment (Fokko) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: My preference would be, making the Operator blocking by default, and also supply the sensor. People can easily use the operator itself if they have only one workbook for example. If they go to more advanced patterns, they can use the sensor pattern. WDYT?

Comment (eladkal) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I have no strong feelings about what should be the default. From my perspective as long as overwrite is possible it's good enough.

Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: @Fokko it really sounds reasonable to me, but I tried to implement it and I don't think it works out nicely... :/

I don't like the two connections one for the refresh operation and one for the sensor task.
If I use the code from the sensor directly - it is duplicated code...

Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: And do I wait until it has succeeded or until it has finished?

**Review by eladkal — APPROVED** [review]
Tested against our Tableau server and it works as expected.

**Comment by feluelle** [comment]
@eladkal please take another look.

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (feluelle) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: I forgot to update this example. If we now have `blocking=True` by default means it will wait until the extract has been refreshed, we don't need the second task.

Or I add `blocking=False`

Comment (Fokko) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: I prefer the first one, but I'll leave it up to you :-)

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (Fokko) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I would `sleep`, similar to here: https://github.com/apache/airflow/blob/aff3a361b4092212c0757f9ce88fa2e40d25d1f4/airflow/providers/apache/druid/hooks/druid.py#L99-L124

File: airflow/providers/salesforce/hooks/tableau.py
Comment (Fokko) [inline airflow/providers/salesforce/hooks/tableau.py:26]: 😍

Comment (feluelle) [reply airflow/providers/salesforce/hooks/tableau.py:26]: 😄

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I am using the Sensor right know with the default mode `poke` means it will `sleep` :)

Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: I need to fully test this behaviour first.

File: setup.py
Comment (feluelle) [reply setup.py]: @Fokko do you think we should move it to a new provider called Tableau?

**Comment by feluelle** [comment]
Ready for another round @Fokko @eladkal. I am going to continue tomorrow :)

Thank you for your time!

**Review by Fokko — APPROVED** [review]
Looking good @feluelle

File: setup.py
Comment (Fokko) [reply setup.py]: Yes, that makes sense to me :)

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (feluelle) [reply airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py:61]: Seems to work :)

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: Shouldn't it be `refresh_tableau_workbook_non_blocking` ?

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: No need for `dag=dag` when you set context manager.

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: Same

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (eladkal) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]: Same

File: airflow/providers/salesforce/hooks/tableau.py
Comment (eladkal) [inline airflow/providers/salesforce/hooks/tableau.py:53]: I asked in previous review about the switching the default of site_id from empty string to None. You had a good point that empty string is the default of the pacakge
https://tableau.github.io/server-client-python/docs/api-ref#authentication
Why did you enetualy decided to switch to None?

File: tests/providers/salesforce/hooks/test_tableau.py
Comment (eladkal) [inline tests/providers/salesforce/hooks/test_tableau.py:33]: Why this connection is required? Can't it use the `tableau_default` that was added to `utils/db.py`?

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (kaxil) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]:
```suggestion
```

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (kaxil) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]:
```suggestion
```

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (kaxil) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]:
```suggestion
```

File: airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py
Comment (kaxil) [inline airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py]:
```suggestion
        job_id="{{ ti.xcom_pull(task_ids='refresh_tableau_workbook_non_blocking') }}",
```

File: airflow/providers/salesforce/hooks/tableau.py
Comment (kaxil) [reply airflow/providers/salesforce/hooks/tableau.py:53]: https://github.com/apache/airflow/pull/7410/files#diff-6332846625a8d68be6a4f2b74a47cc5dR56 takes care of that !

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Good work @feluelle 🎉

**Pull request merged**
